subscriptions/reports.py

```diff
@@ -0,0 +1,252 @@
+        return self.new.count()
+
+    def get_new_datetimes(self) -> List[datetime]:
+        """ List of newly created subscriptions' dates within selected period. """
+        return list(self.new.values_list('start', flat=True))
+
+    @property
+    def ended_or_ending(self) -> QuerySet:
+        """ Subscriptions that end (or gonna end) within selected period. """
+        return (
+            self.overlapping
+            .filter(end__gte=self.since, end__lte=self.until)
+            .ended_or_ending()
+            .order_by('end')
+        )
+
+    def get_ended_count(self) -> int:
+        """ Number of subscriptions ending within selected period. """
+        return self.ended_or_ending.count()
+
+    def get_ended_datetimes(self) -> List[datetime]:
+        """ List of end dates for subscriptions that end within selected period. """
+        return list(self.ended_or_ending.values_list('end', flat=True))
+
+    def get_ended_or_ending_ages(self) -> List[timedelta]:
+        """ List of ages for ended or ending subscriptions."""
+        return self.ended_or_ending.with_ages(at=self.until).values_list('age', flat=True)
+
+    @property
+    def active(self) -> QuerySet:
+        """ Subscriptions that remain active within selected period. """
+        now_ = now()
+        return self.overlapping.exclude(
+            Q(end__lte=now_) | Q(end__gt=now_, auto_prolong=False),
+            end__gte=self.since,
+            end__lte=self.until,
+        ).order_by('start')
+
+    def get_active_count(self) -> int:
+        """ Number of subscriptions that remain active within selected period. """
+        return self.active.count()
+
+    def get_active_users_count(self) -> int:
+        """ Number of users that have active subscriptions within selected period. """
+        return self.active.order_by('user').distinct('user').count()
+
+    def get_active_ages(self) -> List[timedelta]:
+        """ List of ages for active subscriptions. """
+        return self.active.with_ages(at=self.until).values_list('age', flat=True)
+
+    def get_active_plans_and_quantities(self) -> List[Tuple[Plan, int]]:
+        """ List of plan & quantity tuples per subscription. """
+        id_to_plan = {plan.id: plan for plan in Plan.objects.all()}
+        return [
+            (id_to_plan[plan_id], quantity)
+            for plan_id, quantity in self.active.values_list('plan', 'quantity')
+        ]
+
+    def get_active_plans_total(self) -> Counter[Plan]:
+        """ Overall number of quantities per plan. """
+        counter = Counter()
+        for plan, quantity in self.get_active_plans_and_quantities():
+            counter[plan] += quantity
+        return counter
+
+
+@dataclass
+class TransactionsReport(IterPeriodsMixin):
+
+    provider_codename: str
+    since: datetime
+    until: datetime = field(default_factory=now)
+
+    # TODO: some methods will explode when there are multiple currencies
```

Comment:
I was trying to figure out why NO_MONEY has default currency thinking just that :D

Reply:
Here's the story: sometimes there is no data (like no payments during selected period), thus, say, revenue is zero. However, for most reports return type is `Money`, and you cannot just return `int(0)` - cause if you later try to use it in some calculations like `Money(10, 'USD') + 0`, it won't work; and overall returning different types from functions is a bad idea. Same story with `None` - won't work with other `Money()` instances.

So I had to create `NO_MONEY = Money(0, 'USD')`. You can create money without specifying currency like `Money(0)` but it will produce `Money('0', 'XYZ')` which won't work with any other currency at all:
```python
In [4]: Money(0) == Money(0, 'USD')
Out[4]: False

In [5]: Money(0) + Money(0, 'USD')
TypeError: Cannot add or subtract two Money instances with different currencies.
```

`USD` is just most common currency we expect.